scaleway/resource_instance_server.go

```diff
@@ -702,6 +806,99 @@ func resourceScalewayInstanceServerUpdate(ctx context.Context, d *schema.Resourc
 	return append(warnings, resourceScalewayInstanceServerRead(ctx, d, meta)...)
 }
 
+type PrivateNICsHandler struct {
+	ctx            context.Context
+	instanceAPI    *instance.API
+	serverID       string
+	privateNICsMap map[string]*instance.PrivateNIC
+	zone           scw.Zone
+}
+
+func newPrivateNICHandler(ctx context.Context, api *instance.API, server string, zone scw.Zone) (*PrivateNICsHandler, error) {
+	handler := &PrivateNICsHandler{ctx: ctx, instanceAPI: api, serverID: server, zone: zone}
+	return handler, handler.flatPrivateNICs()
+}
+
+func (ph *PrivateNICsHandler) flatPrivateNICs() error {
+	privateNICsMap := make(map[string]*instance.PrivateNIC)
+	res, err := ph.instanceAPI.ListPrivateNICs(&instance.ListPrivateNICsRequest{Zone: ph.zone, ServerID: ph.serverID})
+	if err != nil {
+		return err
+	}
+	for _, p := range res.PrivateNics {
+		privateNICsMap[p.PrivateNetworkID] = p
+	}
+
+	ph.privateNICsMap = privateNICsMap
+	return nil
+}
+
+func (ph *PrivateNICsHandler) detach(o interface{}) error {
+	oPtr := expandStringPtr(o)
+	if oPtr != nil && len(*oPtr) > 0 {
+		// check if old private network still exist on instance server
+		if p, ok := ph.privateNICsMap[expandID(*oPtr)]; ok {
+			// detach private NIC
+			err := ph.instanceAPI.DeletePrivateNIC(&instance.DeletePrivateNICRequest{
```

Comment:
Is it a synchronous operation? Or should a wait be added?

Reply:
It should be added, this check only if the private network was already detached by another process. For example by the console.